planner/core/planbuilder.go

```diff
@@ -1708,19 +1708,24 @@ func (b *PlanBuilder) buildAnalyzeTable(as *ast.AnalyzeTableStmt, opts map[ast.A
 				}
 			}
 			if pruneMode == variable.DynamicOnly || pruneMode == variable.StaticButPrepareDynamic {
-				info := analyzeInfo{DBName: tbl.Schema.O, TableName: tbl.Name.O, TableID: AnalyzeTableID{PersistID: tbl.TableInfo.ID, CollectIDs: physicalIDs}, Incremental: as.Incremental}
-				p.IdxTasks = append(p.IdxTasks, AnalyzeIndexTask{
-					IndexInfo:   idx,
-					analyzeInfo: info,
-					TblInfo:     tbl.TableInfo,
-				})
+				for i, id := range physicalIDs {
+					if id == tbl.TableInfo.ID {
+						id = -1
+					}
+					info := analyzeInfo{DBName: tbl.Schema.O, TableName: tbl.Name.O, PartitionName: names[i], TableID: AnalyzeTableID{TableID: tbl.TableInfo.ID, PartitionID: id}, Incremental: as.Incremental}
+					p.IdxTasks = append(p.IdxTasks, AnalyzeIndexTask{
+						IndexInfo:   idx,
+						analyzeInfo: info,
+						TblInfo:     tbl.TableInfo,
+					})
+				}
 			}
```

Comment:
Assume table `t` has three partitions `p0`, `p1`, `p2`. When executing `analyze table t` in `StaticButPrepareDynamic`, here we generate 6 `IdxTasks` while 3 may be enough since we save partition table stats for dynamic pruning now.